runtime/autoload/ccomplete.vim

```diff
@@ -1,639 +1,8 @@
-      let s:grepCache[a:typename] = qflist
-    endif
-  endif
-
-  " Skip over [...] items
-  let idx = 0
-  while 1
-    if idx >= len(a:items)
-      let target = ''		" No further items, matching all members
-      break
-    endif
-    if a:items[idx][0] != '['
-      let target = a:items[idx]
-      break
-    endif
-    let idx += 1
-  endwhile
-  " Put matching members in matches[].
-  let matches = []
-  for l in qflist
-    let memb = matchstr(l['text'], '[^\t]*')
-    if memb =~ '^' . target
-      " Skip matches local to another file.
-      if match(l['text'], "\tfile:") < 0 || bufnr('%') == bufnr(matchstr(l['text'], '\t\zs[^\t]*'))
-	let item = {'match': memb, 'tagline': l['text']}
-
-	" Add the kind of item.
-	let s = matchstr(l['text'], '\t\(kind:\)\=\zs\S\ze\(\t\|$\)')
-	if s != ''
-	  let item['kind'] = s
-	  if s == 'f'
-	    let item['match'] = memb . '('
-	  endif
-	endif
-
-	call add(matches, item)
-      endif
-    endif
-  endfor
-
-  if len(matches) > 0
-    " Skip over next [...] items
-    let idx += 1
-    while 1
-      if idx >= len(a:items)
-	return matches		" No further items, return the result.
-      endif
-      if a:items[idx][0] != '['
-	break
-      endif
-      let idx += 1
-    endwhile
-
-    " More items following.  For each of the possible members find the
-    " matching following members.
-    return s:SearchMembers(matches, a:items[idx :], a:all)
-  endif
-
-  " Failed to find anything.
-  return []
-endfunc
-
-" For matching members, find matches for following items.
-" When "all" is non-zero find all, otherwise just return 1 if there is any
-" member.
-func s:SearchMembers(matches, items, all)
-  let res = []
-  for i in range(len(a:matches))
-    let typename = ''
-    if has_key(a:matches[i], 'dict')
-      if has_key(a:matches[i].dict, 'typename')
-	let typename = a:matches[i].dict['typename']
-      elseif has_key(a:matches[i].dict, 'typeref')
-	let typename = a:matches[i].dict['typeref']
-      endif
-      let line = "\t" . a:matches[i].dict['cmd']
-    else
-      let line = a:matches[i]['tagline']
-      let e = matchend(line, '\ttypename:')
-      if e < 0
-	let e = matchend(line, '\ttyperef:')
-      endif
-      if e > 0
-	" Use typename field
-	let typename = matchstr(line, '[^\t]*', e)
-      endif
-    endif
-
-    if typename != ''
-      call extend(res, s:StructMembers(typename, a:items, a:all))
-    else
-      " Use the search command (the declaration itself).
-      let s = match(line, '\t\zs/^')
-      if s > 0
-	let e = match(line, '\<' . a:matches[i]['match'] . '\>', s)
-	if e > 0
-	  call extend(res, s:Nextitem(strpart(line, s, e - s), a:items, 0, a:all))
-	endif
-      endif
-    endif
-    if a:all == 0 && len(res) > 0
-      break
-    endif
-  endfor
-  return res
-endfunc
-
-let &cpo = s:cpo_save
-unlet s:cpo_save
-
-" vim: noet sw=2 sts=2
+" Generated vim file by vim9jit. Please do not edit
+let s:PATH = expand("<script>")
```

Comment:
We currently use uppercase for `vim.log.levels.[TRACE,DEBUG,INFO,...]`, and there's also the constant I added [I added ](https://github.com/neovim/neovim/blob/c3d8665851c20414007f04a481c5a542d9d139f2/runtime/lua/vim/_editor.lua#L321).

Using uppercase for constants is just a simple signal that a variable isn't intended to be mutated and I think it's a particularly useful convention for scripting languages like python, lua and shell.

Reply:
but is it really? Why is it so important to SIGNAL THIS EVERYWHERE (sorry, caps :) compared to all the numerous other aspects a variable can have?  If immutability is important it should be enforced mechanically, not through this cargo-cult.

Reply:
> to all the numerous other aspects a variable can have?

Like what? Type and scope? Scope isn't as important if the thing is a constant. Type can be determined via `type()` since it's an intrinsic property to the variable.

>  If immutability is important it should be enforced mechanically

This isn't reasonably possible in Lua, and if there's a way it has heavy costs if something like metatables are used.

I said this is only useful in scripting languages where constants can't be enforced. For languages like C(++), rust, scala, java, or even Lua 5.4 (which has `<const>`), this convention doesn't give you anything.

It's an inoffensive convention with no downsides (unless caps are offensive), so I don't understand the complaint. I've also seen camelcase used for modules variables (`local Util = require'plugin.util'`). Whilst I've not used this myself, I can see its appeal. Would you also be against that? Must all variables be lower-snake_case for absolute consistency?

Reply:
Uppercase *global* variables are saved to shada

Reply:
done

Reply:
> It's an inoffensive convention with no downsides (unless caps are offensive), so I don't understand the complaint.

 complaint is that it's very inconsistently and randomly used (cargo-cult) and doesn't actually have a purpose. 

> isn't reasonably possible in Lua, and if there's a way it has heavy costs if something like metatables are used.

To enforce immutability, returning a value from a function seems reasonable.

> I've also seen camelcase used for modules variables (local Util = require'plugin.util')

I'm not against conventions :) But the uppercase const convention feels like an "emperor no clothes" situation.

Reply:
i could lock the vars ;) but the file is generated by vim9jit, so it doesn't matter and should never be hand edited (it will generated all the available vimscript functions for a particular autoload file)

Reply:
> complaint is that it's very inconsistently and randomly used (cargo-cult)

How exactly is it inconsistent and randomly used? 

> ... and doesn't actually have a purpose.

As mentioned, it's just a signal to the reader that the variable is a simple value and doesn't represent state. Take the 3 cases:

```lua
for i = 0, 20 do ... end         -- 1
for i = 0, some_max do ... end   -- 2
for i = 0, SOME_MAX do           -- 3
```

1 should be avoided since magic numbers are bad. 

2 gives no indication whether `some_max` is expected to change (and LSP can't even tell you). 

3 indicates the loop is fixed and doesn't depend on any state.

So overall I find this pretty value and certainly makes refactoring easier.

> To enforce immutability, returning a value from a function seems reasonable.

- A function can indicate it might perform side effects or return a different value each time.
- Is much more tedious to write and reference
- Is slower and harder for a compiler to optimize

Immutability isn't the thing that's particularly important. It's more about stateful vs stateless code.

Reply:
It's not necessarily about being readonly, more about whether something is a static constant (does it depend on state?). Admittedly the case with `PATH` is a bit grey.

Reply:
i'll let you sort this out later and I can update the generated code later ;)